sympy/ntheory/qs.py

```diff
@@ -0,0 +1,487 @@
+        """
+        ans = 0
+        for coeff in self.modified_coeff:
+            ans *= x
+            ans += coeff
+        return ans
+
+class factor_base_elem:
+    """This class stores an element of the factor_base.
+    """
+    def __init__(self, prime, tmem_p, log_p):
+        """
+        Initialization of factor_base_elem.
+
+        Parameters:
+        ===========
+
+        prime : prime number of the factor_base
+        tmem_p : Integer square root of x**2 = n mod prime
+        log_p : Compute Natural Logarithm of the prime
+        """
+        self.prime = prime
+        self.tmem_p = tmem_p
+        self.log_p = log_p
+        self.soln1 = None
+        self.soln2 = None
+        self.a_inv = None
+        self.b_ainv = None
+
+
+def _generate_factor_base(num_primes, n):
+    """Generate factor_base for Quadratic Sieve. The factor_base
+    consists of all the the points whose legendre_symbol(n, p) == 1
+    and p < num_primes. Along with the prime factor_basealso stores
+    natural logarithm of prime and the residue n modulo p.
+
+    Parameters:
+    ===========
+
+    num_primes : upper bound of the factor_base
+    n : integer to be factored
+    """
+    from sympy import prime
+    factor_base = []
+    i = 1
+    while len(factor_base) < num_primes:
+        ith_prime = prime(i)
+        i += 1
+        if pow(n, (ith_prime - 1) // 2, ith_prime) == 1:
+            residue = _sqrt_mod_prime_power(n, ith_prime, 1)[0]
+            log_p = round(log(ith_prime)*2**10)
+            factor_base.append(factor_base_elem(ith_prime, residue, log_p))
+    return factor_base
+
+def _initialize_first_polynomial(N, M, factor_base):
+    """This step is the initialization of the 1st sieve polynomial.
+    Here `a` is selected as a product of several primes of the factor_base
+    such that `a` is about to sqrt(2*N) / M. Other initial values of
+    factor_base elem are also intialized which includes a_inv, b_ainv, soln1,
+    soln2 which are used when the sieve polynomial is changed. The b_ainv
+    is required for fast polynomial change as we don't have to calculate
+    `2*b*mod_inverse(a, prime)` every time.
+
+    Parameters:
+    ===========
+
+    N : Number to be factored
+    M : sieve interval
+    factor_base : factor_base primes
+    """
+    approx_val = sqrt(2*N) / M
+    # `a` is a parameter of the sieve polynomial and `q` is the prime factors of `a`
+    # randomly search for a combination of primes whose multiplication is close to approx_val
+    # This multiplication of primes will be `a` and the primes will be `q`
+    # `best_a` denotes that `a` is close to approx_val in the random search of combination
+    best_a, best_q, best_ratio = None, None, None
+    for _ in range(50):
+        a = 1
+        q = []
+        while(a < approx_val):
+            rand_p = 0
+            while(rand_p == 0 or rand_p in q):
+                rand_p = random.randint(3, len(factor_base) - 1)
```

Comment:
@jksuom  This is a major thing that i wanted to discuss. in  both the paper and https://www.rieselprime.de/ziki/Self-initializing_quadratic_sieve there is a bound for       factors. In        the paper only lower bound is set to 2000. While in the link there is a bound that the factor should in (2000, 4000) range. Which       one is better, and i think a bound  should be necessary cause too small factors increases the chance of getting redundant relations.

Reply:
I think that the lower bound is set to make it more efficient to create a product of size about `sqrt(2N)/M`. The upper bound is probably not critical, it could be bigger if N is big. Maybe some experimentation is necessary.

Reply:
Yea i will experiment with this. Also i think that the current process of selection of `a` in `initialize_first_poly` might       not be most efficient. I will search about this.

Reply:
> I think that the lower bound is set to make it more efficient to create a product of size about `sqrt(2N)/M`. The upper bound is probably not critical, it could be bigger if N is big. Maybe some experimentation is necessary.

I think that the upper bound is important. Suppose in initialization stage we use 10 primes to generate `a` then we have 2**10 -1 polynomials to use. But if the upper bound is not set then most probably                    we             will generate less number of polynomials per `a` then we have to select another `a`.